### DISCUSSION TITLE
State of the Union

### DISCUSSION THREAD

**Discussion by cmxog** [tact-lang/spec#4] [General]
Currently, the way `union` syntax is defined in the parser makes it very similar to other things like fns, structs, both in full and all forms of shorthands:

```
union A { B, C}
union A(T: Type) { B(T), C}
let A = union { B, C};
let A(T: Type) = union { B(T), C}
```
and so on.

There has been a suggestion to change the syntax of unions to something like `B | C` for brevity and perhaps to demote them from the fully "typeship".

I've thought about it a lot and I believe that the pros of this (brevity, for the most part) do not outweigh the cons, which I believe to be:

* By looking at the `B | C` syntax, even if one can guess the meaning (Rust, Haskell, TypeScript developers will be; others may be not so much), one won't necessarily know how we call it. Call it `union` in syntax makes it extremely obvious as to what it is.
* The language loses its consistency and therefore, discoverability. Having a `let` or inline-only availability of union syntax violates the principle of the least surprise (that is, all types follow a similar pattern)
* This syntax does not have the space for defining methods as the currently implemented one. I know there have been an opinion that perhaps unions should not have member methods, but I am not 100% sure that's the right call. Rust's enums do. In C++'s `std::variant`, at the very least you can inherit from a specific variant and add methods to your liking.

I know there's an argument that `|` is more visual (and to those that have used this syntax for type definitions, it is useful) but we also agreed that the use of punctuation to introduce meanings is something we should be careful about.

So, my current proposal is to keep the direction as we did them originally. Retain the syntax and afford them full typeship in the sense that they can have methods.

Thoughts, comments?

**Solution**: see [comment](https://github.com/tact-lang/docs/discussions/4#discussioncomment-2863806).

**Comment by p0lunin** [comment]
Pros of the `|` syntax:
1. It more neat and smaller than `union {...}`.
2. It does not confuse users, who may think that unions create newtypes which is wrong in Tact.

> By looking at the B | C syntax, even if one can guess the meaning (Rust, Haskell, TypeScript developers will be; others may be not so much), one won't necessarily know how we call it. Call it union in syntax makes it extremely obvious as to what it is.

This isn't a problem as I think. `||` often used as logical OR in many languages so `|` is intuitive `OR` statement.

> The language loses its consistency and therefore, discoverability. Having a let or inline-only availability of union syntax violates the principle of the least surprise (that is, all types follow a similar pattern)

As I wrote above, our unions does not create newtypes, unlike `struct`, `interface` or `enum`. So if we add a `union {}` syntax that does not match user expectations, we lose in consistency, not otherwise.

> This syntax does not have the space for defining methods as the currently implemented one.

At my opinion unions must implement methods if all members of the unions implement it like in the follow fragment:
```
let Some(T) = type{ T; fn unwrap(self) -> T { return self.T } }; // weird syntax, yes
let None = type{ fn unwrap(self) -> T { /* fail */ } };
let Option(T) = Some(T) | Nothing;
fn do_stuff() {
  Some{10}.unwrap(); // 10
  None{}.unwrap(); // fail
}
```
In this case user do not need to:
1. Override methods on the union (see [trait_enum](https://docs.rs/trait_enum/latest/trait_enum/) and [enum_dispatch](https://docs.rs/enum_dispatch/latest/enum_dispatch/) crates in Rust which solve this problem using macros).
2. Any methods implemented on the enums (not unions) basically starts with matching, see [std::option](https://doc.rust-lang.org/src/core/option.rs.html#530) in Rust.

**Reply by oleganza** [reply]
I agree with the most of your arguments, and will comment on the issue separately. 

Just one note on the idea of propagating method impls: that might be a brittle decision similar to Go's implied interface impls. I suggest making this rule for the interfaces only:

> Union of types {A,B,...} implement interfaces X,Y,... if every type in the union implements all of these interfaces.

**Comment by oleganza** [comment]
I'd like to zoom out a bit and enumerate some facts first.

1. We want sum-types ("unions/enums") and product-types ("structs").
2. We need to choose between _nominal_ typing and _structural_ typing.
3. Rust/Swift offer nominal types for both enums and structs.
4. Haskell/TypeScript offer structural types for both enums and structs.
5. In Tact we are trying nominal types for structs ("point{x,y} != vector{x,y}") and structural for unions (less boilerplate than in Rust). 
6. Nominal typing requires a concept of "new type" that needs well-defined support in our compile-time language ("how to distinguish creation of new type VS instantiation of another copy of the same type?").

Questions:

1. Is (5) correct choice? Could there be a problem with such inconsistency, despite the declared benefits?
2. Are there languages that choose similar asymmetry? (Well, TS/JS has nominal classes VS structural objects/unions.)
3. Are there languages that make the opposite choice: nominal unions and structural product-types?

Now, back to the syntax question within current paradigm. Here are my thoughts on this:

1) Syntax `union X { A, B }` looks very symmetrical with `struct X {a:, b: }` while with different newtype semantics that need to be explained over and over to people who at first will find it odd. This also violates the principle of the least surprise.

2) Syntax `A | B` clearly reads "A or B" by virtually any programmer, and that construct will be found in almost every program, so there will be very little discoverability issues. We are unlikely to reuse this syntax for bitwise ops (those could be covered by explicitly named functions), based on the principle "punctuation for fundamental language features only".

3) We might imagine a built-in function `union(A,B)`, but for a fundamental feature this would look quite silly.

4) Yes, `|`-union does not have a "nice" shorthand as `struct X { ... }`: it only has anonymous form and must be assigned via `let`. However, it's by design: since we explicitly do not want a newtype for the union, it only makes sense to keep it anonymous.

5) There probably will be another bit of punctuation operator for types: `+` (or `&`, but plus sign looks cleaner) for interfaces. Say we want to declare `let x: Readable+Writeable`. So there will be two fundamental punctuations: `|` for type disjunction and `+` for type conjunction (electrical engineers roll their eyes at this point since `+` is used for binary `OR`).

**Comment by oleganza** [comment]
In the view of the problems with semantic and syntactic asymmetry of `struct A` and `|`, I have some fun proposal:

1. We do structural typing by default, throughout. 
2. `struct{ }` generates anonymous structurally-typed struct.
3. We keep `|` syntax.
4. `struct T {...}` reserves `T` in the namespace (assuring one-time use) and adds it as a Zero-Sized tag on the struct, enforcing uniqueness when comparing `Vector{x,y}` with `Point{x,y}`. Also creates binding `T`.
5. Repeated use of `struct T{...}` performs standard structural checks to ensure that T binding is not re-bound to a different type.

Therefore:

```
let P = struct{x:T, y:T}  // creates anonymous struct {x:T,y:T} that is the same wherever
let P = struct Point{x:T, y:T} // creates newtype struct "Point".
let P = struct P {} // fails because P binding is already created by struct, so `let P` won't work.
```

Generics and pattern matching:

This is no good:

```
let P = fn(T: Type){ struct Point{x: T, y: T} };
...
let P1 = P(Int); 
let P2 = P(Int);
```

In the above example, in P2, struct is created second time, attempts to re-bind to "Point" and attempt succeeds because the struct is the same. But it won't work if we did `P(Int)` first and then `P(String)`.

Better version:

```
let P = fn(T: Type){ struct Point#T {x: T, y: T} };
...
let P1 = Point(Int); 
let P2 = Point(String);
```

Here we construct the newtype tag using the argument type explicitly with some extra crazy syntax. However, that's a footgun akin to missing dependencies in React: what if you don't mix in all the arguments that matter and not let some uses downstream?

We either should make newtype structs quasi-functions with arguments, or define special memoization behavior as a part of the language. I'd try to offer the former.

**Reply by oleganza** [reply]
The above proposal is ugly because we allow mixing newtypes and anonymous types, which may cause a lot of confusion down the road:

```
struct Point{x:,y:}
...
let P1 = struct{x:,y:}
let P2 = struct{x:,y:}
```

If P1 == P2, then users may also expect that Point == P1, but that wouldn't be the case, although visually that does not follow. If `struct Point` has explicit field `_tag: "Point"`, then the difference would be clear, but that'd be ugly as fuck.

I guess if we want nominal structs, we need to have ONLY nominal structs. But then maybe let X = struct{} should be reconsidered and we need something explicitly parametric: `struct X(T) { ... }` (not as a sugar, but at the fundamental level).

**Comment by cmxog** [comment]
I think part of the confusion here is that there are unions and there are tagged unions, at least on the conceptual level. Realistically, we are probably looking for tagged union one way or another (am I wrong?).

One issue that came up in my mind today is the integration of our type system with the serialization framework. If we can't define methods on unions, how do we tag them in serialization? Let's say there's a `Serializable` interface (details are omitted for brevity):

```
interface Serializable {
  fn serialize(self: Self)
}
```

In this conversation there was an idea that a union implements an interface if and only if all of its members implement it. However, in this particular case we'll always serialize the member without serializing its tag, which is likely to be a problem when we'll need to deserialize it as we won't be able to tell what that type is. 

There is a workaround for this, that requires a discipline (and I don't favor it). We can stipulate that all types should be record with some identifying information. But that becomes murky as types are rather complex (parametric, etc.). Whereas if we want to encode member tag at a union level, that is as simple as putting a member index down.

So, if we have kept the current approach to unions (where they *can* have methods; or at the very least, implementations of interfaces, which are not very different from arbitrary methods anyway), we can simply implement `Serializable` on that union which would record a member index and call its member's `serialize` method.

**Reply by p0lunin** [reply]
Implementing `serialize` method directly on the type is a VERY BAD idea.

Suppose we have 2 foreign actors: `Receiver1` and `Receiver2`. Suppose both of them accept message `Accept { value: Int257, data: Data1 | Data2 }`, but one of them has ABI `accept_message$0 value:Int257 data:Data = Accept` while other has ABI `accept_message$0 data:Data value:Int257 = Accept`. So in this case we need _two_ serialization methods that should have _two_ different method names corresponding to the each actor.

That's what about `serialize` method. What about implementing methods - we can remember that we have strong compile-time system so we can define function `extend(t: Type, fn_name: Ident, fn: Function)` that will add new methods on the `t` type as in followed example:
```
let Option(T) = Some(T) | Nothing;
extend(Option, unwrap, fn(self: Option(T)) -> T { ... })
```
Yes, this isn't ergonomic way. To be more ergonomic I proposed annotations syntax:
```
@extend(Option)
fn unwrap(T)(self: Option(T)) -> T { ... }
```

**Comment by cmxog** [comment]
Based on our discussion with @p0lunin, there's another proposal for the syntax for union that both preserves the terminological clarity (gives a term to use) and the brevity of the pipe (`|`):

```
union Variant | Varariant_ | ...
```

This is also easier to parse as there won't be a need to disambiguate a single expression and a start of a union. The `union` keyword would clearly indicate what is it to parse there.

It can perhaps be allowed to be extended to have an optional leading pipe:

```
let T = union
| A
| B;
```

If we want it to have a consistent shorthand form (similar to `struct`, `fn`, etc.) it maybe still be not great (but better than `A|B`) as it'd result in something like

```
union T A | B
```

But it is not necessarily bad, and it is consistent with the general shorthand form (but not the curly brackets, simply because we don't have them in this version of union's syntax). It may get a bit weird if we use a leading pipe in this syntax:

```
union T
| A
| B
```

while this looks good, this is a bit harder to read (while being the same AST):

```
union T | A | B
```

which also looks like a non-shorthand version of union of three members.

This proposal likely depends on Tact using [Rust-inspired out-of-band method declarations (`impl`)](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2717067), unless we will allow it to be extended to have methods as well:

```
let T = union A | B {
  ... methods ...
};
```

then we can perhaps use the [consolidated model](https://github.com/tact-lang/spec/discussions/6#discussioncomment-2717092) as well?

**Reply by cmxog** [reply]
I was thinking a bit more about this syntax, and if we do want to have shorthand consistency, I would suggest using `union [name] of [variant] | [variant]`. The overhead is minimal but the meaning is clear and the parsing is straightforward.

**Comment by p0lunin** [comment]
Resulting union syntax is as follows:
```ebnf
union_def = "union", "{", case, { case }, methods, "}";
case = "case", expr, [ ";" ];
sugared_union_def = "union", ident, [generics], "{", case, { case }, methods, "}";
```

This syntax chosen due to following reasons:
1. It is consistent with other syntax defintions: `struct` and `interface`.
2. It is easely to add methods to the union with this syntax.

While first reason isn't important, second is more important. With `A | B` syntax there are no explicit possibility to add methods to an union, which forces us to add new syntax or semantic constructions to allow this. We do not want to add such constructions if there are way to not do this.
